### PR TITLE
Overload array expressions/patterns to also work for floatarray

### PR DESCRIPTION
**Pull request by nojb**
This PR proposes to overload array expressions `[| e1; e2; ...; eN |]` and the corresponding patterns so that they can be used for both `float array` and `floatarray`. The overloading is decided based on the "expected type" (coming from the context of the expression or pattern), which means that in certain situation an explicit annotation `(... : floatarray)` may be needed to force the disambiguation.

A few examples are included to see the effect.

Opinions/suggestions welcome!

### CODE REVIEW

**Comment by nojb** [comment]
cc @lpw25 @gasche

**Comment by Drup** [comment]
I'm not sure overloaded constants is a pandora box we want to open casually. Constants of type `_ format` are already fairly problematic for this reason.

If we start going that direction, we need a global decision with a decent plan on how it's going to integrate everywhere else, not just a hack for `Floatarray.t`.

**Comment by alainfrisch** [comment]
Well, it's not really about constants, more about the array "constructor".  Overloading it using the expected type does not look so different from usual overloading of sum-type constructor.  It's just that currently we don't have a syntax to define other "array types".

**Comment by lpw25** [comment]
I'm in favour of the feature, although I think it should include the array projections too. However, I'm pretty keen on the implementation using the same disambiguation code as variant constructors and record labels, as it does in #616, rather than their own ad-hoc code. Without using those code paths I don't think the behaviour for e.g. principality warnings, warning 40 etc. will be quite right.

Roughly speaking I would expect this to mean instantiating the `NameChoice` functor with a `lookup_all_from_type` that just checked whether the `Path.t` was either `array` or `floatarray`.

**Comment by nojb** [comment]
> Roughly speaking I would expect this to mean instantiating the `NameChoice` functor with a `lookup_all_from_type` that just checked whether the `Path.t` was either `array` or `floatarray`.

Thanks for the hint, I will give it a try!

**Comment by Drup** [comment]
My issue is that, as usual, we have a bunch of features that do not exactly align. We already have a feature for array projections: operator redefinitions: `let (.[]) = ....`. @lpw25 is very keen on not doing that for actual arrays to keep space for this upcoming-since-2016 PR.

So, either we sit down and decide that array creation is a data-constructors and we use type-based disambiguation, or we decide that array creation is like function application, and we just add new operator for arrays (like `let ([|;|]) = ...`), in the line of other array operators. Regardless of the decision, the array PR need to fall in line with it if we are ever to merge it.

In any case, 1) it shouldn't be an exclusive hack for floatarray 2) it's a language-design decision that is not so inconsequential. 3) as @lpw25 pointed out, the current implementation is not right.

**Comment by nojb** [comment]
cc @yallop

**Comment by nojb** [comment]
> So, either we sit down and decide that array creation is a data-constructors and we use type-based disambiguation, or we decide that array creation is like function application, and we just add new operator for arrays (like `let ([|;|]) = ...`), in the line of other array operators. Regardless of the decision, the array PR need to fall in line with it if we are ever to merge it.

The "function application" viewpoint would not solve the problem of pattern-matching, right? With that in mind, the "data constructor" viewpoint seems strictly better.

**Comment by gasche** [comment]
One argument in favor of seeing `[| ... |]` as a container-type-overloadable construction would be that `[ ... ]` already is (by virtue of being syntactic sugar for variant constructors).

**Comment by nojb** [comment]
I pushed a commit making use of the `NameChoice` machinery, but as far as I can see it does not bring that much to the table as we always default to `'a array` in the absence of constraints...

**Comment by nojb** [comment]
Also pushed one more commit extending the toplevel printer to print `floatarray` values as well.

**Comment by lpw25** [comment]
> but as far as I can see it does not bring that much to the table

It should cause principality warnings and warnings 41 and 42 to be emitted when appropriate. You should add a test case for each of the `warn_foo` calls in `NameChoice(_).disambiguate` to make sure it is all working properly -- excluding the two underneath the call to `lookup_from_type` since I would expect those to be unreachable.

I also think your dummy implementation is a little too dummy -- I'll add a couple of comments about that in the code.

There do not appear to be any changes to the typed tree or the transl functions in lambda. Which means this patch can't be generating the right code in the float array case in no-flat-float-array mode. That needs to be fixed.

File: typing/typecore.ml [hunk 1]
```diff
@@ -845,6 +862,40 @@ let wrap_disambiguate msg ty f x =
   | Wrong_name_disambiguation (env, wrong_name) ->
     raise (Error (wrong_name.name.loc, env, Wrong_name (msg, ty, wrong_name)))
 
+(* Use a dummy instance of [NameChoice] to disambiguate between [float array]
+   and [floatarray]. Note that an actual ambiguous choice is not possible, as we
+   default to ['a array] in case of doubt. *)
+
+module FloatArray = NameChoice(struct
+    type t = bool
```
Comment (lpw25) [inline typing/typecore.ml]: I think this type should be expanded to be more like `label_description`. It should include the type of the array elements and the type of the resulting array as type expressions, alongside whether or not it is a float array. It would also be good to use a variant rather than a bool for that part.

This will allow you to implement `get_type` correctly rather than using `Predef.type_unit`. It will also allow you to use these types directly in `Ppat_array` and `Pexp_array` rather than computing them as `ty_elt` and `to_unify`.

File: typing/typecore.ml [hunk 2]
```diff
@@ -845,6 +862,40 @@ let wrap_disambiguate msg ty f x =
...
+    type usage = unit
+    let kind = Datatype_kind.Record
+    let get_name = function
```
Comment (lpw25) [inline typing/typecore.ml]: I think this function should just be the constant function returning something like `"([||])"`. It should match the longident that is passed into disambiguate. If you expand the patch to disambiguate array projection as well -- which I think you should -- then using `"(.())"` would make the most sense.

**Comment by nojb** [comment]
> There do not appear to be any changes to the typed tree or the transl functions in lambda. Which means this patch can't be generating the right code in the float array case in no-flat-float-array mode. That needs to be fixed.

As far as I can see (and with some cursory testing), this works already because `Typeopt.array_type_kind` returns the right thing.

File: typing/typecore.ml [hunk 3]
```diff
@@ -845,6 +862,40 @@ let wrap_disambiguate msg ty f x =
...
+      | false -> Predef.type_array Predef.type_unit
+      | true -> Predef.type_floatarray
+    let lookup_all_from_type _loc () p _env =
+      [Path.same p Predef.path_floatarray, ignore]
+    let in_env _ = true
+  end)
+
+let disambiguate_array_type loc env expected_ty =
+  let lid = Location.mkloc (Longident.Lident "#array#") loc in
+  let expected_type, candidates =
```
Comment (lpw25) [inline typing/typecore.ml]: The candidates should always be the equivalent of `Ok [false, ignore; true, ignore]`, they should not depend on whether there is an expected type.

File: typing/typecore.ml [hunk 4]
```diff
@@ -845,6 +862,40 @@ let wrap_disambiguate msg ty f x =
...
+    let get_type = function
+      | false -> Predef.type_array Predef.type_unit
+      | true -> Predef.type_floatarray
+    let lookup_all_from_type _loc () p _env =
+      [Path.same p Predef.path_floatarray, ignore]
+    let in_env _ = true
+  end)
+
+let disambiguate_array_type loc env expected_ty =
+  let lid = Location.mkloc (Longident.Lident "#array#") loc in
```
Comment (lpw25) [inline typing/typecore.ml]: `"#array#"` should be replaced with whatever `get_name` returns.

**Comment by lpw25** [comment]
> this works already because Typeopt.array_type_kind returns the right thing

Urgh. Please don't rely on the type-based optimisation code to "optimise" the generated code from incorrect to correct.

**Comment by nojb** [comment]
Thank you @lpw25 for the detailed explanation.

I pushed a new rewritten version of the patch addressing your comments. The tests show the new warnings that we get thanks to using `NameChoice`.

I disabled warning 41 for array constructors as per our discussion on Slack.

As for the matter of overlading the `.(...)` notation, my sense is that this part is less consensual or at least hasn't been discussed as much. I can add that to this PR if there is consensus, but otherwise maybe it is better left for later. Opinions?

**Comment by nojb** [comment]
Friendly ping. I'm keen to move this PR forward as it has a substantial impact on the usability of the `floatarray` type.

**Comment by gasche** [comment]
What is the behavior of the proposal in absence of type information? Do we get an error, or a warning, or do we implicitly make the non-principal choice of defaulting to the `array` type?

If I understand correctly, currently the last behavior (implicit default to `array`) is implemented. Is this the behavior we want? For compatibility reasons?

(I think this should be clarified in the PR description; ideally you would include example of programs with annotations, with propagated type information, and with no type information, to see what choices are made and what the user is told.)

File: testsuite/tests/typing-extensions/floatarray.ml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  | _ -> ()
+;;
+[%%expect{|
+val f : Float.Array.t -> unit = <fun>
+|}, Principal{|
+Line 4, characters 4-8:
+4 |   | [||] -> ()
+        ^^^^
+Warning 18 [not-principal]: this type-based field disambiguation is not principal.
+val f : Float.Array.t -> unit = <fun>
```
Comment (gasche) [inline testsuite/tests/typing-extensions/floatarray.ml:70]: The warning here is a bit odd, given that this typing is in fact principal (there is no other valid type than `floatarray` for this pattern), and that we do *not* have a warning in the actually-not-principal case `let x = [| 1.2 |]` (no type annotations, `array` chosen by default).

Comment (lpw25) [reply testsuite/tests/typing-extensions/floatarray.ml:70]: The warning is the same as we get with record disambiguation. It's a bit of a stretch, but I think of it as being not principal between the two options of "types as `floatarray`" and "gives a type error".

**Comment by gasche** [comment]
I don't like the non-principal default to `'a array` in absence of type information. I think we should have a non-principal warning there, when the element type is `float`.

It might be too inconvenient in practice, with warnings popping up in existing code, but then

1. Maybe the warning wouldn't actually show up that often, so it would be fine.
2. Maybe users of float-array code could actually take such warnings as an encouragement to migrate their code to `Floatarray`, so that would be good.

File: testsuite/tests/typing-extensions/floatarray.ml (hunk 5 above)
Comment (gasche) [reply testsuite/tests/typing-extensions/floatarray.ml:70]: Note: also "field disambiguation" is wrong in the error message, maybe the functor would need to be made more parametric on the sort of disambiguation.

@lpw25 the difference I make with variant/record disambiguation is that, for variants, you also get a warning when you construct variants in a non-principal way, whereas here the warning shows up in cases that are actually principal, and there is no warning in cases that are not.

**Comment by nojb** [comment]
> I don't like the non-principal default to `'a array` in absence of type information. I think we should have a non-principal warning there, when the element type is `float`.

Specifically, I believe this would be Warning 41 ("Ambiguous constructor or label name"). This is currently disabled in the code for array constructors (together with a comment to that effect). As I understand from some discussion with @lpw25, the issue is that constructor disambiguation cannot take into account the type of the arguments (here, the type of array elements) because this is not available yet, as the type of the constructor is used to obtain the "expected type" of the arguments.

> It might be too inconvenient in practice, with warnings popping up in existing code, but then
> 
> 1. Maybe the warning wouldn't actually show up that often, so it would be fine.
> 2. Maybe users of float-array code could actually take such warnings as an encouragement to migrate their code to `Floatarray`, so that would be good.

**Comment by gasche** [comment]
That's a good point, but I still think that we should have a non-principality warning if we are not sure that the code is principal. Then if this warning proves too annoying, we can think of refinement to the code (in the functor or outside it) to be more clever in determining principality.

File: testsuite/tests/typing-extensions/floatarray.ml (hunk 5 above)
Comment (lpw25) [reply testsuite/tests/typing-extensions/floatarray.ml:70]: > for variants, you also get a warning when you construct variants in a non-principal way

I don't think that you do:
```ocaml
# type t = Foo;;
type t = Foo
# type s = Foo;;
type s = Foo
# let x = Foo;;
val x : s = Foo
```
at least not by default. I don't think anyone actually turns on warning 41.

Comment (lpw25) [reply testsuite/tests/typing-extensions/floatarray.ml:70]: Also worth noting:
```
# module M = struct
    type t = Foo
  end;;
module M : sig type t = Foo end

# type s = Foo;;
type s = Foo

# #warnings "+41";;

# let x = Foo;;
val x : s = Foo

# let x : M.t = Foo;;
val x : M.t = M.Foo
```

Comment (gasche) [reply testsuite/tests/typing-extensions/floatarray.ml:70]: Erm, you are correct. So the behavior in this PR is as nonsensical as the behavior already observed for variants. I'll drop my criticism of this particular testsuite example, but I remain of the opinion that we should get a non-principality warning when we write non-principal code. (And I think that doing this properly in this PR may require a code modification.)

Note: in retrospect it's not completely true that the behavior is the same, there is a difference between "no warning whatsoever" and "no warning by default". Anyway.

**Comment by lpw25** [comment]
If we are sticking to the notion of disambiguation as used for records and variants then I think there are two options that are consistent:
1. Consider the floatarray version of `[| |]` to be part of the initial environment. That means that any use of `[| |]` without a type annotation would emit warning 41 (if anyone has it enabled).
2. Consider the floatarray version of `[| |]` to not be part of the initial environment. That means that any use of the floatarray version of `[| |]` would emit warning 40 (again if anyone has it enabled).

My original proposal took route 2. Although you could also avoid warning 40 by opening `Float.Array` since that re-exported the floatarray `[| |]` constructor.

The current patch does 1 but then disables warning 41 for array constructors. This may also be fine but its not quite consistent with records and variants.

**Comment by gasche** [comment]
For a reminder for people who forgot about those warnings:

> 18 Non-principal type.
> [...]
> 40 Constructor or label name used out of scope.
> 41 Ambiguous constructor or label name.

In my mind it is pretty clear that scoping is a separate issue from principality, and that it shouldn't get into play here for a built-in language construct. I would expect a default-disambiguated array literal to raise warnings 18 *and* 41. If the current implementation only raises 41 and not 18, then maybe Leo's proposal to follow that is reasonable, but at least the warning should be 41 and not 40.

**Comment by jberdine** [comment]
Can I ask what the intended behavior with respect to principality and ambiguity is for the empty array literal `[||]`? What about functions such as `let one x = [|x|]`? Are these cases that are currently polymorphic and unambiguous expected to be deemed non-principal and ambiguous since there is no element type information? If so, then I feel that it would be a step backward.

(And just as a data point, we have code that compiles in principal mode and has warnings 40 and 41 enabled.)

**Comment by gasche** [comment]
Well now `let one x = [|x|]` may mean either `let one x : floatarray = [|x|]` or `let one x : _ array = [|x|]`, and the two are different. So yes, the code is not principal and you should get a warning, *especially* if you compile with `-principal` and warning `41` enabled. When you say that "it would be a step backward", are you complaining about a warning here, or about the feature that makes the code non-principal?

(It's not so much about the element type than about the array type. We discussed the element type because I proposed that the cases where the element type is known to be non-float could be automatically deduced to be `array`.)

**Comment by jberdine** [comment]
Yes, I appreciate that technically once `floatarray` expressions are added to the language, then these cases become non-principal. What isn't clear to me is whether the intent it to promote the `floatarray` type to such first-class status, in a way that introduces ambiguity in code that was previously unambiguous.

I'm probably overlooking several things, but is it correct that there are basically two sorts of usage for which good support is desired:
  1. Code that does not consciously use floatarrays, and essentially ignores their existence; and
  2. Code that explicitly uses floatarrays.

I expect that most code of the first sort would rather essentially ignore the existence of floatarrays, which ambiguity and principality warnings would prevent. For the second sort, is it unreasonable to expect `Float.Array` to be opened? While it may not be technically entirely uniform, it seems to me that it would be better for existing code and code that does not use floatarrays if the `[|` `|]` expressions and patterns referred only to `array` unless `Float.Array` is opened, at which point the code is asking for both to be considered, and the resulting ambiguity warnings are intended. To my understanding, this supports @lpw25's approach 2 above.

I am not overly worried about this non-principality leading to situations where a code change causes disambiguation to make a different choice "far" away, causing programmer confusion about how their change is at fault for the type error. I'm more worried about needing to switch principal mode off due to what will be deemed "spurious" warnings.

As an aside, it is potentially curious to overload `[||]` when other values such as `0` and `0.` and `0L` need to have their types resolved by using distinct syntax. Is it problematic to use e.g. `[.| |.]` or something for floatarrays instead?

**Comment by nojb** [comment]
> Is it problematic to use e.g. `[.| |.]` or something for floatarrays instead?

Technically this would be perfectly acceptable. In fact, it would be simpler for the user (no annotations needed, ever), and simpler to implement (no need for fancy disambiguation). Actually, this is the first thing that came to my mind. But I guess there is something of an inertia *against* adding new syntax among many of the developers, and an approach based on disambiguation seemed to get more traction :)

Anyway, I am happy to prototype this approach as well, and I will try to do so if I get the time.

**Comment by Drup** [comment]
@nojb That was the alternative design I mentioned. We allow user-defined operator for `[XX|....|]` or `[XX|....|XX]` in the style of indexation operator. You will even be able to do `let open Floatarray.Op in` to get the vanilla array syntax if you want.

And maybe later, once we have proper ad-hoc polymorphism we can add poly-shape array operators that do not compromise principality.

**Comment by nojb** [comment]
> @nojb That was the alternative design I mentioned. We allow user-defined operator for `[XX|....|]` or `[XX|....|XX]` in the style of indexation operator. You will even be able to do `let open Floatarray.Op in` to get the vanilla array syntax if you want.

This would work for array construction, but not for pattern matching. An ad-hoc syntax would allow for both, but would be, well, ad-hoc :)

**Comment by Drup** [comment]
Sorry if that question is naive, but is there any actual use case of pattern matching fix-length arrays? And even worse: *float* arrays?

**Comment by nojb** [comment]
> Sorry if that question is naive, but is there any actual use case of pattern matching fix-length arrays? And even worse: _float_ arrays?

No, the question is not naive. In theory one would imagine there isn't much use for it, but in practice pattern matching gives a very convenient readable way to name the elements of an array in numeric code. For example, the following is taken frm LexiFi's code base:
```ocaml
let multiply_vector a b =
  match a, b with
  | [|a11;a21;a31;a41;
      a12;a22;a32;a42;
      a13;a23;a33;a43;
      a14;a24;a34;a44|],
    [|b1;b2;b3;b4|] ->
    [|
      a11*.b1 +. a12*.b2 +. a13*.b3 +. a14*.b4;
      a21*.b1 +. a22*.b2 +. a23*.b3 +. a24*.b4;
      a31*.b1 +. a32*.b2 +. a33*.b3 +. a34*.b4;
      a41*.b1 +. a42*.b2 +. a43*.b3 +. a44*.b4; |]
  | _ -> failwith "multiply: dim not supported"
```

**Comment by lpw25** [comment]
> Well now let one x = [|x|] may mean either let one x : floatarray = [|x|] or let one x : _ array = [|x|], and the two are different. So yes, the code is not principal and you should get a warning, especially if you compile with -principal and warning 41 enabled. 

I do wish people would stop saying that this code is not principal. It is fine from a principality perspective: it's principal type is `'a -> 'a array`. Principality is not about whether things can be typed without annotations, it is about the relationship between the inference algorithm and the declarative specification. It is perfectly reasonable to define the declarative specification to require annotations in certain positions.

An example of a non-principal piece of code would be:
```ocaml
let one x =
  if p then ([| 3.5 |] : Float.Array.t)
  else [|x|]
```
This is where the `-principal` warning should get raised. (Technically, you could produce a declarative specification that allowed this too, but only by sacrificing other properties of the declarative specification: the further you go down this route the less declarative your specification becomes).

The original code is however ambiguous from the perspective of warning 41, which is the actual point of the discussion.

As for using a separate operator, I come back to the point: why treat arrays differently from other data types? All of the issues being discussed in this thread also apply to variant constructors and record fields.

I think that, based on @jberdine's reasoning, the best approach is to:
1. Change the patch so that the environment keeps track of which array constructors are in scope.
2. Only have the normal array constructor in scope in the initial environment
3. Make it so that `open Float.Array` brings the other array constructor into scope

My preferred way to achieve 3 is a little involved, but it is conceptually simple and will naturally scale to the full array data types proposal:
1. Add a `Type_array` constructor to `type_kind` to represent a type which introduces array constructors to the environment
2. Add an extension point to represent this new type kind
3. Use the extension point in `Float.Array`'s reexport of `float_array`

**Comment by gasche** [comment]
I would like to get back on this question of principality. I believe that, with array overloading, we lose principality.

I wrote:
>  Well now `let one x = [|x|]` may mean either `let one x : floatarray = [|x|]` or `let one x : _ array = [|x|]`, and the two are different [none of them is an instance of the other]. So yes, the code is not principal and you should get a warning, especially if you compile with -principal and warning 41 enabled.

@lpw25 wrote:
> I do wish people would stop saying that this code is not principal. It is fine from a principality perspective: it's principal type is `'a -> 'a array`. Principality is not about whether things can be typed without annotations, it is about the relationship between the inference algorithm and the declarative specification. It is perfectly reasonable to define the declarative specification to require annotations in certain positions.

Okay, so what is the declarative specification that you have in mind as the typing rule for `[|...|]` ?

From my point of view, there are two declarative typing rules:

```
e1, .., en : t
-----------------------------
[| e1; ...; en |] : t array
```

and 

```
e1, .., en : float
-----------------------------
[| e1; ...; en |] : Floatarray.t
```

and the program `let one x = [|x|]` admits two typing derivations, one with `one : 'a -> 'a array` and one with `one : float -> Floatarray.t`. The fact that none of the derivations is "more general" than the other means that we lost principality.

(The second typing rule may be "available" all the time, or only when `Floatarray` was opened in lexical scope, but in either proposals there exist regions of the program where the two typing rules are available concurrently.)

If we had a typing rule that requires an annotation to use the second rule, for example `([| ... |] : Floatarray.t)`, then we would be principal again. (But the case of type-disambiguated constructors and fields, which also has the same problem, would remain principality-breaking.) But then we may as well consider an extension syntax, `[|...|]%float` for example.

**Comment by gasche** [comment]
I would like to clarify that I think it may very well be okay to lose principality, if we judge that the ergonomics of programming are better without it. But if this is the case, I believe that (1) it should be acknowledged, and (2) it should be possible to enable a warning to see the places where principality is lost in this way (and, in particular, I would expect `-principal` to reject the unannotated program).

@lpw25 pointed out during private discussion that other OCaml features break principality. One example he gave is polymorphic methods. I would be interested in seeing a concrete example spelled out in details (allowing users to explicitly go outside the ML fragment with annotations does not necessarily break principality, as long as the unannotated program fails to type-check), but I can believe that this is indeed the case. But note that any non-principal typing there is associated with an operation `m#f` which is known not to have annotation-only polymorphic behavior. (For `m.f` it is a bit harder to argue; in absence of disambiguation we can say that the behavior is principally enforced by the type declaration of the type `f`.) In any case, the existence of some principality-breaking feature is not by itself a justification to introduce other principality-breaking features without discussing it and trying to avoid it.

**Comment by lpw25** [comment]
I don't think anyone has bothered to write out the principal typing rules for constructor disambiguation -- which would correspond to array construction -- but the rules for record label disambiguation seem similar enough to polymorphic methods that I can borrow the rules from [Semi-explicit First-class Polymorphism](https://www.sciencedirect.com/science/article/pii/S0890540199928304) to outline some typing rules for array projection quite easily.

Extending the `[σ]ᵋ` construct to allow `[floatarray]ᵋ` then I think the key rules would be something like:
```
  Γ ⊢ e : ∀ε.[floatarray]ᵋ    Γ ⊢ i : int
 ―――――――――――――――――――――――――――――――――――――――――
  Γ ⊢ e.(i) : float

  Γ ⊢ e : float array    Γ ⊢ i : int
 ――――――――――――――――――――――――――――――――――――
  Γ ⊢ e.(i) : float
```
alongside a rule for float array annotations:
```
  Γ ⊢ e : floatarray
 ――――――――――――――――――――――――――――――――――――――
  Γ ⊢ (e : floatarray) : [floatarray]ᵋ
```

This is pretty close to the actual implementation of `-principal`  in the type-checker.

> @lpw25 pointed out during private discussion that other OCaml features break principality.

No, I pointed out that other features of OCaml have different valid types in the presence or absence of type annotations. As I keep telling you, this *does not mean they break principality*. It just means that the notion of type in the system is more complicated than you suppose.

Here is a simple example:
```ocaml
# let foo x y z = x#m y + x#m z;;
val foo : < m : 'a -> int; .. > -> 'a -> 'a -> int = <fun>

# let foo (x : < m : 'a. 'a -> int>) y z = x#m y + x#m z;;
val foo : < m : 'a. 'a -> int > -> 'b -> 'c -> int = <fun>
```

The other places that might cause you to raise warning 18 may also allow such examples: GADTs, incomplete subtyping coercions, first-class modules, optional arguments, commuting labelled arguments.

As a side-note, my understanding is that there are some corner cases where OCaml does properly break principality. The point is that the cases that you are referring to, and to which this PR adds one more, are not those corners.

**Comment by gasche** [comment]
Privately @trefis proposed another approach, which is to use a type system that incorporates "expected type" information, and restrict the floatarray-literal introduction rule to the case where the expected type is known to be `Floatarray.t` (expected types mostly appear thanks to annotations).

Leo's proposal on first inspection looks like a somewhat more complex way to say the same thing. In the Semi-explicit polymorphism work, there is the idea first that some program constructs explicitly eliminate polymorphic types (method calls, field projections), and then that to play nicely with inference those polymorphic types should not be "guessed", but forced to come from user-written annotations. "Epsilon variables" are used to shoe-horn annotation-propagation within the usual discipline of ML inference (generalization, instantiation, etc.).

I don't know what is the connection today between epsilon-variable propagation and expected-type propagation in the OCaml type system (does one strategy subsume the others, do they have interesting differences?), but in this particular case we are discussing I believe that expected-type propagation is a more natural way to look at it, and in fact it is already the approach used for type-disambiguated constructors and fields.

To summarize: Thomas and Leo are proposing a different type system specification where the `[| ... |] : floatarray` typing rule is not allowed in general, but only when it is placed under a user annotation. It is only with this restriction that the proposed system can be seen as principal.

**Comment by Drup** [comment]
> Privately @trefis proposed another approach, which is to use a type system that incorporates "expected type" information, and restrict the floatarray-literal introduction rule to the case where the expected type is known to be `Floatarray.t` (expected types mostly appear thanks to annotations).

If I remember correctly, that's what is done for format strings.
